core/dbt/clients/registry.py

```diff
@@ -74,15 +76,58 @@ def package(name, registry_base_url=None):
             use_name = response["name"]
 
         new_nwo = use_namespace + "/" + use_name
-        deprecations.warn("package-redirect", old_name=name, new_name=new_nwo)
+        deprecations.warn("package-redirect", old_name=package_name, new_name=new_nwo)
+
+    return response
+
+
+_get_cached = memoized(_get)
+
+
+def package(package_name, registry_base_url=None):
+    # returns a dictionary of metadata for all versions of a package
+    response = _get_with_retries(package_name, registry_base_url)
+    return response["versions"]
+
+
+def package_version(package_name, version, registry_base_url=None):
+    # returns the metadata of a specific version of a package
+    response = package(package_name, registry_base_url)
+    return response[version]
+
+
+def get_available_versions(package_name):
+    # returns a list of all available versions of a package
+    response = package(package_name)
+    return list(response)
+
+
+def _get_index(registry_base_url=None):
+
+    url = _get_url("index", registry_base_url)
```

Comment:
I like these functions and how they abstract over retries. Can we add type hints to them?